### ISSUE TITLE
New (group14/16, sha256/512) kexen server mode broken

### ISSUE THREAD

**Issue by bitprophet** [paramiko/paramiko#1462]
A user confirms that they are experiencing this problem for both "diffie-hellman-group16-sha512" and "diffie-hellman-group14-sha256" KEX algorithms in server mode. The client is 'SSH-2.0-1.82 sshlib: ClientSftp'

They later confirmed it's reproducible with OpenSSH-client. See below link & subsequent comments.

_Originally posted by @miverson in https://github.com/paramiko/paramiko/issues/1455#issuecomment-503372188_

**Comment by miverson** [comment]
As the reporting user, I'd be happy to assist in any way possible.

**Comment by miverson** [comment]
Is there an ETA on this fix entering a production release?

**Comment by kuwv** [comment]
@miverson Check the milestones.

**Comment by ploxiln** [comment]
fixed in #1883

**Comment by miverson** [comment]
Has this been fixed in 2.7.2? 

I seem to be getting the same error.

Client side:
```
➜  sftp -o KexAlgorithms=diffie-hellman-group16-sha512 
ssh_dispatch_run_fatal: Connection to xx port 22: incorrect signature
Connection closed
➜  sftp -o KexAlgorithms=diffie-hellman-group14-sha256 
ssh_dispatch_run_fatal: Connection to xx port 22: incorrect signature
Connection closed
➜  ssh -V
OpenSSH_8.1p1, LibreSSL 2.7.3
```

Server side:
```
sshd.py:sshd.py:50 INFO     Accepted connection from ip xx
sshd.py:transport.py:1819 DEBUG    starting thread (server mode): 0xf2444fa0
sshd.py:transport.py:1819 DEBUG    Local version/idstring: SSH-2.0-paramiko_2.7.2
sshd.py:transport.py:1819 DEBUG    Remote version/idstring: SSH-2.0-OpenSSH_8.1
sshd.py:transport.py:1819 INFO     Connected (version 2.0, client OpenSSH_8.1)
sshd.py:transport.py:1819 DEBUG    kex algos:['diffie-hellman-group16-sha512', 'ext-info-c'] server key:['ecdsa-sha2-nistp256-cert-v01@openssh.com', 'ecdsa-sha2-nistp384-cert-v01@openssh.com', 'ecdsa-sha2-nistp521-cert-v01@openssh.com', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384', 'ecdsa-sha2-nistp521', 'ssh-ed25519-cert-v01@openssh.com', 'rsa-sha2-512-cert-v01@openssh.com', 'rsa-sha2-256-cert-v01@openssh.com', 'ssh-rsa-cert-v01@openssh.com', 'ssh-ed25519', 'rsa-sha2-512', 'rsa-sha2-256', 'ssh-rsa'] client encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] server encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] client mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] server mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] client compress:['none', 'zlib@openssh.com', 'zlib'] server compress:['none', 'zlib@openssh.com', 'zlib'] client lang:[''] server lang:[''] kex follows?False
sshd.py:transport.py:1819 DEBUG    Kex agreed: diffie-hellman-group16-sha512
sshd.py:transport.py:1819 DEBUG    HostKey agreed: ecdsa-sha2-nistp256
sshd.py:transport.py:1819 DEBUG    Cipher agreed: aes128-ctr
sshd.py:transport.py:1819 DEBUG    MAC agreed: hmac-sha2-256-etm@openssh.com
sshd.py:transport.py:1819 DEBUG    Compression agreed: none
sshd.py:transport.py:1819 DEBUG    kex engine KexGroup16SHA512 specified hash_algo <built-in function openssl_sha512>
sshd.py:transport.py:1819 ERROR    Socket exception: Connection reset by peer (104)
----------------------------------------
Exception occurred during processing of request from ('', 59658)
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/socketserver.py", line 616, in process_request
    self.finish_request(request, client_address)
  File "/usr/local/lib/python3.9/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python3.9/socketserver.py", line 747, in __init__
    self.handle()
  File "/opt/sftp/sshd.py", line 78, in handle
    self.transport.start_server(server=srv_iface)
  File "/usr/local/lib/python3.9/site-packages/paramiko/transport.py", line 728, in start_server
    raise e
  File "/usr/local/lib/python3.9/site-packages/paramiko/transport.py", line 2055, in run
    ptype, m = self.packetizer.read_message()
  File "/usr/local/lib/python3.9/site-packages/paramiko/packet.py", line 459, in read_message
    header = self.read_all(self.__block_size_in, check_rekey=True)
  File "/usr/local/lib/python3.9/site-packages/paramiko/packet.py", line 301, in read_all
    x = self.__socket.recv(n)
ConnectionResetError: [Errno 104] Connection reset by peer
----------------------------------------
sshd.py:sshd.py:50 INFO     Accepted connection from ip xxx
sshd.py:transport.py:1819 DEBUG    starting thread (server mode): 0xf2404400
sshd.py:transport.py:1819 DEBUG    Local version/idstring: SSH-2.0-paramiko_2.7.2
sshd.py:transport.py:1819 DEBUG    Remote version/idstring: SSH-2.0-OpenSSH_8.1
sshd.py:transport.py:1819 INFO     Connected (version 2.0, client OpenSSH_8.1)
sshd.py:transport.py:1819 DEBUG    kex algos:['diffie-hellman-group14-sha256', 'ext-info-c'] server key:['ecdsa-sha2-nistp256-cert-v01@openssh.com', 'ecdsa-sha2-nistp384-cert-v01@openssh.com', 'ecdsa-sha2-nistp521-cert-v01@openssh.com', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384', 'ecdsa-sha2-nistp521', 'ssh-ed25519-cert-v01@openssh.com', 'rsa-sha2-512-cert-v01@openssh.com', 'rsa-sha2-256-cert-v01@openssh.com', 'ssh-rsa-cert-v01@openssh.com', 'ssh-ed25519', 'rsa-sha2-512', 'rsa-sha2-256', 'ssh-rsa'] client encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] server encrypt:['chacha20-poly1305@openssh.com', 'aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-gcm@openssh.com', 'aes256-gcm@openssh.com'] client mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] server mac:['umac-64-etm@openssh.com', 'umac-128-etm@openssh.com', 'hmac-sha2-256-etm@openssh.com', 'hmac-sha2-512-etm@openssh.com', 'hmac-sha1-etm@openssh.com', 'umac-64@openssh.com', 'umac-128@openssh.com', 'hmac-sha2-256', 'hmac-sha2-512', 'hmac-sha1'] client compress:['none', 'zlib@openssh.com', 'zlib'] server compress:['none', 'zlib@openssh.com', 'zlib'] client lang:[''] server lang:[''] kex follows?False
sshd.py:transport.py:1819 DEBUG    Kex agreed: diffie-hellman-group14-sha256
sshd.py:transport.py:1819 DEBUG    HostKey agreed: ecdsa-sha2-nistp256
sshd.py:transport.py:1819 DEBUG    Cipher agreed: aes128-ctr
sshd.py:transport.py:1819 DEBUG    MAC agreed: hmac-sha2-256-etm@openssh.com
sshd.py:transport.py:1819 DEBUG    Compression agreed: none
sshd.py:transport.py:1819 DEBUG    kex engine KexGroup14SHA256 specified hash_algo <built-in function openssl_sha256>
sshd.py:transport.py:1819 ERROR    Socket exception: Connection reset by peer (104)
----------------------------------------
Exception occurred during processing of request from ('', 59663)
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/socketserver.py", line 616, in process_request
    self.finish_request(request, client_address)
  File "/usr/local/lib/python3.9/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python3.9/socketserver.py", line 747, in __init__
    self.handle()
  File "/opt/sftp/sshd.py", line 78, in handle
    self.transport.start_server(server=srv_iface)
  File "/usr/local/lib/python3.9/site-packages/paramiko/transport.py", line 728, in start_server
    raise e
  File "/usr/local/lib/python3.9/site-packages/paramiko/transport.py", line 2055, in run
    ptype, m = self.packetizer.read_message()
  File "/usr/local/lib/python3.9/site-packages/paramiko/packet.py", line 459, in read_message
    header = self.read_all(self.__block_size_in, check_rekey=True)
  File "/usr/local/lib/python3.9/site-packages/paramiko/packet.py", line 301, in read_all
    x = self.__socket.recv(n)
ConnectionResetError: [Errno 104] Connection reset by peer
----------------------------------------
```

**Comment by bitprophet** [comment]
@miverson No, this will be 2.7.3 or whatever goes out the door next. Hopefully soon.
